Enumio should use name() instead of toString()

danielb987 opened the pull request:
In some enum classes, toString() is used for other things than returning the name of the enum, for example to return a localized name. So name() should be used instead of toString() when reading/writing xml files.

This PR fixes that and adds a check for it by adding the method toString() to the test enum clas.

@bobjacobsen 
If `AbstractXmlAdapter.EnumIoNames.inputFromString()` gets a string that's not a valid enum, it returns a `null`. Shouldn't this be an error? If I have the enum FooBar with the values Foo and Bar, and I `call inputFromString("ooF")`, I get a `null` as result and a log message with level `trace`.

I think that if the element exists in the xml file but the string is wrong, it should be an error which results in a IllegalArgumentException or at least an error message to the log. If an unknown string is in the xml file, it's probably a program error or a xml error, like in my case there I have toString() overridden.

bobjacobsen commented:
I agree that `toString()` and `name()` don't always say the same thing.

I picked `toString()` exactly because of that: it provided a way to vary what's stored/loaded in the file vs what programmers see.  For example, if you want to rename the `MAN_I_SHOULD_HAVE_PICKED_A_SHORTER_NAME` enum item because you're tired of typing it, using `toString()` would have provided a way to keep storing and loading the same file.

But that was before EnumIO had the two-sided map implementation, which is a better solution.  So I'm fine with using `name()`.

As to error handing: the XML Schema is supposed to be the first line of defense of that.  I'm not sure a `Throwable` is the right solution, because it'll really mess up the read process since it's not likely to be caught low down.  EnumIO should probably do an error report before it does anything else though.

danielb987 commented:
Thanks for your answer. I have added an error message if the enum cannot be found.

> As to error handing: the XML Schema is supposed to be the first line of defense of that. I'm not sure a `Throwable` is the right solution, because it'll really mess up the read process since it's not likely to be caught low down. EnumIO should probably do an error report before it does anything else though.

If `inputFromString()` cannot find the enum, it returns `null`. But it's marked as returning `@Nonnull`. So it breaks its own contract. Should `@Nonnull` be removed?

The parameter to `inputFromString()` is also marked as `@Nonnull`. Which means that this code is not enough since there is no check for null:
```
AbstractXmlAdapter.EnumIO<LnStringIO.Type> map =
        new AbstractXmlAdapter.EnumIoNames<>(LnStringIO.Type.class);
lnStringIO.setType(map.inputFromString(e.getChildText("type")));
```
If the method `inputFromString()` can return `null`, I think that it would also be okay to take `null` as the parameter. But if you prefer to have the parameter as non null, I will add the null check.

The same questions apply to `inputFromAttribute()` which also has a `@Nonnull` parameter and a `@Nonnull` result.

danielb987 commented:
Spotbugs failed due to:
`[ERROR] jmri.configurexml.AbstractXmlAdapter$EnumIoNames.inputFromString(String) may return null, but is declared @Nonnull [jmri.configurexml.AbstractXmlAdapter$EnumIoNames, jmri.configurexml.AbstractXmlAdapter$EnumIoNames, jmri.configurexml.AbstractXmlAdapter$EnumIoNames] Returned at AbstractXmlAdapter.java:[line 329]Null value at AbstractXmlAdapter.java:[line 324]Known null at AbstractXmlAdapter.java:[line 325] NP_NONNULL_RETURN_VIOLATION`

So this needs to be fixed, either by removing @Nonnull or by throwing an exception if null.

danielb987 commented:
Would it be good to add a method `inputFromString(String s, T default)` which returns a default value if null?

danielb987 commented:
This PR should probably be replaced by #8557.

danielb987 commented:
I have updated this PR with the https://github.com/JMRI/JMRI/pull/8557#issuecomment-631856570 of @bobjacobsen in #8557.

`EnumIO.inputFromString()` now accepts null and if the enum cannot be found or an exception is thrown, inputFromString() returns the first enum of the enum type. And I have added tests for this.

danielb987 commented:
@bobjacobsen 
If you think this PR solves the issues you wrote about in https://github.com/JMRI/JMRI/pull/8557#issuecomment-631856570, this PR is ready to be merged. Unless you want to wait for next production release which is fine for me.

bobjacobsen commented:
Thanks!
